From: Andrey Rakhmatullin <wrar@debian.org>
Date: Sun, 23 Mar 2025 16:20:16 +0500
Subject: Update some requirement versions to sid ones.

---
 src/Cargo.toml                                     | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -21,11 +21,11 @@
 license = "Apache-2.0 OR BSD-3-Clause"
 
 [workspace.dependencies]
-asn1 = { version = "0.22.0", default-features = false }
-pyo3 = { version = "0.26", features = ["abi3"] }
-pyo3-build-config = { version = "0.26" }
-openssl = "0.10.74"
-openssl-sys = "0.9.110"
+asn1 = { version = "0.20.0", default-features = false }
+pyo3 = { version = "0.22.6", features = ["abi3-py313"] }
+pyo3-build-config = { version = "0.22.6" }
+openssl = "0.10.65"
+openssl-sys = "0.9.103"
 
 [profile.release]
 overflow-checks = true
--- a/src/rust/cryptography-cffi/Cargo.toml
+++ b/src/rust/cryptography-cffi/Cargo.toml
@@ -12,7 +12,7 @@
 openssl-sys.workspace = true
 
 [build-dependencies]
-cc = "1.2.37"
+cc = "1.2.19"
 
 [lints.rust]
 unexpected_cfgs = { level = "warn", check-cfg = ['cfg(python_implementation, values("CPython", "PyPy"))'] }
