--- a/src/cryptography/hazmat/backends/openssl/backend.py
+++ b/src/cryptography/hazmat/backends/openssl/backend.py
@@ -289,13 +289,15 @@
         if self._lib.CRYPTOGRAPHY_NEEDS_OSRANDOM_ENGINE:
             # Unregister and free the current engine.
             self.activate_builtin_random()
-            with self._get_osurandom_engine() as e:
-                # Set the engine as the default RAND provider.
-                res = self._lib.ENGINE_set_default_RAND(e)
+
+            if self._lib.FIPS_mode() == 0:
+                with self._get_osurandom_engine() as e:
+                    # Set the engine as the default RAND provider.
+                    res = self._lib.ENGINE_set_default_RAND(e)
+                    self.openssl_assert(res == 1)
+                # Reset the RNG to use the engine
+                res = self._lib.RAND_set_rand_method(self._ffi.NULL)
                 self.openssl_assert(res == 1)
-            # Reset the RNG to use the engine
-            res = self._lib.RAND_set_rand_method(self._ffi.NULL)
-            self.openssl_assert(res == 1)
 
     def osrandom_engine_implementation(self):
         buf = self._ffi.new("char[]", 64)
