diff --git a/cryptography/hazmat/bindings/openssl/ssl.py b/cryptography/hazmat/bindings/openssl/ssl.py
index 7d805e7..139a5dd 100644
--- a/cryptography/hazmat/bindings/openssl/ssl.py
+++ b/cryptography/hazmat/bindings/openssl/ssl.py
@@ -390,8 +390,12 @@ const long SSL_OP_LEGACY_SERVER_CONNECT = 0;
 SSL_METHOD* (*SSLv2_client_method)(void) = NULL;
 SSL_METHOD* (*SSLv2_server_method)(void) = NULL;
 #else
+#ifdef OPENSSL_NO_SSL2
+static const long Cryptography_HAS_SSL2 = 0;
+#else
 static const long Cryptography_HAS_SSL2 = 1;
 #endif
+#endif
 
 #ifdef SSL_CTRL_SET_TLSEXT_HOSTNAME
 static const long Cryptography_HAS_TLSEXT_HOSTNAME = 1;
