diff --git a/src/_cffi_src/openssl/src/osrandom_engine.c b/src/_cffi_src/openssl/src/osrandom_engine.c
index 52f55af..6492e36 100644
--- a/src/_cffi_src/openssl/src/osrandom_engine.c
+++ b/src/_cffi_src/openssl/src/osrandom_engine.c
@@ -271,7 +271,7 @@ static int osrandom_init(ENGINE *e) {
                     CRYPTOGRAPHY_OSRANDOM_R_GETRANDOM_INIT_FAILED_EAGAIN,
                     __FILE__, __LINE__
                 );
-                getrandom_works = CRYPTOGRAPHY_OSRANDOM_GETRANDOM_INIT_FAILED;
+                getrandom_works = CRYPTOGRAPHY_OSRANDOM_GETRANDOM_INIT_FAILED_EAGAIN;
                 break;
             default:
                 /* EINTR cannot occur for buflen < 256. */
@@ -299,7 +299,14 @@ static int osrandom_init(ENGINE *e) {
 static int osrandom_rand_bytes(unsigned char *buffer, int size) {
     long n;
 
+    if (getrandom_works == CRYPTOGRAPHY_OSRANDOM_GETRANDOM_INIT_FAILED_EAGAIN) {
+        /* Try to initialize again -- the kernel CPRNG may now be seeded */
+        (void) osrandom_init(NULL);
+    }
+
     switch(getrandom_works) {
+    case CRYPTOGRAPHY_OSRANDOM_GETRANDOM_INIT_FAILED_EAGAIN:
+        /* fall through */
     case CRYPTOGRAPHY_OSRANDOM_GETRANDOM_INIT_FAILED:
         ERR_Cryptography_OSRandom_error(
             CRYPTOGRAPHY_OSRANDOM_F_RAND_BYTES,
@@ -345,6 +352,8 @@ static int osrandom_finish(ENGINE *e) {
 
 static int osrandom_rand_status(void) {
     switch(getrandom_works) {
+    case CRYPTOGRAPHY_OSRANDOM_GETRANDOM_INIT_FAILED_EAGAIN:
+        return 0;
     case CRYPTOGRAPHY_OSRANDOM_GETRANDOM_INIT_FAILED:
         return 0;
     case CRYPTOGRAPHY_OSRANDOM_GETRANDOM_NOT_INIT:
@@ -359,6 +368,8 @@ static int osrandom_rand_status(void) {
 
 static const char *osurandom_get_implementation(void) {
     switch(getrandom_works) {
+    case CRYPTOGRAPHY_OSRANDOM_GETRANDOM_INIT_FAILED_EAGAIN:
+        return "<failed>";
     case CRYPTOGRAPHY_OSRANDOM_GETRANDOM_INIT_FAILED:
         return "<failed>";
     case CRYPTOGRAPHY_OSRANDOM_GETRANDOM_NOT_INIT:
diff --git a/src/_cffi_src/openssl/src/osrandom_engine.h b/src/_cffi_src/openssl/src/osrandom_engine.h
index d28ebf3..581287e 100644
--- a/src/_cffi_src/openssl/src/osrandom_engine.h
+++ b/src/_cffi_src/openssl/src/osrandom_engine.h
@@ -52,7 +52,8 @@
 #endif
 
 enum {
-    CRYPTOGRAPHY_OSRANDOM_GETRANDOM_INIT_FAILED = -2,
+    CRYPTOGRAPHY_OSRANDOM_GETRANDOM_INIT_FAILED_EAGAIN = -3,
+    CRYPTOGRAPHY_OSRANDOM_GETRANDOM_INIT_FAILED,
     CRYPTOGRAPHY_OSRANDOM_GETRANDOM_NOT_INIT,
     CRYPTOGRAPHY_OSRANDOM_GETRANDOM_FALLBACK,
     CRYPTOGRAPHY_OSRANDOM_GETRANDOM_WORKS
