FIPS: make test RSA keys SP800-56b compatible
diff --git a/tests/hazmat/primitives/fixtures_rsa.py b/tests/hazmat/primitives/fixtures_rsa.py
index 09b32ab00..1c4ad0db7 100644
--- a/tests/hazmat/primitives/fixtures_rsa.py
+++ b/tests/hazmat/primitives/fixtures_rsa.py
@@ -590,14 +590,14 @@ RSA_KEY_2048 = RSAPrivateNumbers(
         16,
     ),
     d=int(
-        "aef17f80f2653bc30539f26dd4c82ed6abc1d1b53bc0abcdbee47e9a8ab433abde865"
-        "9fcfae1244d22de6ad333c95aee7d47f30b6815065ac3322744d3ea75058002cd1b29"
-        "3141ee2a6dc682342432707080071bd2131d6262cab07871c28aa5238b87173fb78c3"
-        "7f9c7bcd18c12e8971bb77fd9fa3e0792fec18d8d9bed0b03ba02b263606f24dbace1"
-        "c8263ce2802a769a090e993fd49abc50c3d3c78c29bee2de0c98055d2f102f1c5684b"
-        "8dddee611d5205392d8e8dd61a15bf44680972a87f040a611a149271eeb2573f8bf6f"
-        "627dfa70e77def2ee6584914fa0290e041349ea0999cdff3e493365885b906cbcf195"
-        "843345809a85098cca90fea014a21",
+        "4e340141b6c1956dd7e5da52ca030ed8afe7410112e39ab54bf9289c29d0cbb43191c"
+        "807f148c6208524aabdd8c5077c2bafc079968c5087ee15743aab300467c9ddd83f68"
+        "ffcfc93ed648f9b57f3de0a03a3bb3a91f5bed67f4cd9d9e46be810fb4aaa3a0e784a"
+        "6a04fce63d09d8e5fae4ab86ecf0ccd73108b8673e6be8496c35631d7fcb65cba63bb"
+        "240538b153bc68443d3532af6376de438b8fab96874b833384e0c45eb41e7ccb20177"
+        "5af03f7190eb55c068798f495610c00c7f4d08c288b7b15fd27786e5776b16f3da0ed"
+        "237115ca250100079e42f6763716a5eada0a3491033c60bba4e6199315dade787866a"
+        "1f1df842672209b5ce5463ca788f9",
         16,
     ),
     dmp1=int(
